src/core/control/Control.cpp

```diff
@@ -1779,7 +1779,28 @@ void Control::undoRedoPageChanged(PageRef page) {
     }
 }
 
-void Control::selectTool(ToolType type) {
+void Control::handleSelectToolAction(ToolType type) {
+    // keep text-selection when switching from text to seletion tool
+    auto oldTool = getToolHandler()->getActiveTool();
+    if (oldTool && win
+                && (type == TOOL_SELECT_RECT || type == TOOL_SELECT_REGION || type == TOOL_SELECT_OBJECT)
+                && oldTool->getToolType() == ToolType::TOOL_TEXT
+                && !(this->win->getXournal()->getTextEditor()->getText()->getText().empty())) {
+        auto xournal = this->win->getXournal();
+        Text* textobj = std::move(xournal->getTextEditor()->getText());
+        clearSelectionEndText();
+
+        auto pageNr = getCurrentPageNo();
+        XojPageView* view = xournal->getViewFor(pageNr);
+        if (view == nullptr) {
+            return;
+        }
```

Comment:
Maybe I'm missing something, but under what scenario is `view` null? My understanding is that every page should have a view; if not, there is something wrong going on. Also, if `view` is indeed null, the `return` statement will prevent the tool switch from happening. 

I would replace this with an assertion:
```suggestion
        g_assert(view != nullptr);
```